### PR TITLE
Adding new interaction contexts

### PR DESCRIPTION
**Pull request by ian-leggett**
## Problem
We need to add new interaction contexts which will control what service options are available to the user depending on what type of interaction they are adding.
Trello - https://trello.com/c/QjT2K2I3/643-add-import-export-other-option-to-interaction

## Solution
Update the initial create interaction form (create kind) with new options that post the new contexts to the next stage of the form. By posting the new contexts in step 1 the "Service" drop down options on the next step of the form process becomes clearer and easier to consume as the list is now populated based on the context you provided.

### Before
Screenshot 2019-05-01 at 10 02 51

### After
Screenshot 2019-05-01 at 10 03 42

interaction-screenshot

### As a result the service options are effected on the next page
Screenshot 2019-04-18 at 14 48 31

### Contacts
1. Add Export interaction via contacts
2. You should see the "Service" context options on the next page
3. Add Export service delivery via contacts
4. You should see the "Service" context options on the next page
5. Add Investment interaction via contacts
6. You should see the "Service" context options on the next page
7. Add Other interaction via contacts
8. You should see the "Service" context options on the next page
9. Add Other service delivery via contacts
10. You should see the "Service" context options on the next page
11 Do 1 - 10 through companies
12 Add an interaction via investments and you should go straight through to the form and see the "Service" context options on the next page

Note: To check what the service context options are log into Django and look at services.

### New contexts to be added

```
export_interaction
export_service_delivery
investment_interaction
other_interaction
other_service_delivery
```

**Implementation notes**

_Add any additional information about the change that isn't captured in the Trello ticket._

**Checklist**

- [ ] Has the branch been rebased to develop?
- [x] Automated tests (Any of the following when applicable: Unit, Functional or Acceptance)
- [x] Manual compatibility testing (Browsers: Chrome, Firefox, IE11, Safari)

### CODE REVIEW

**Comment by jim68000** [comment]
This pull request **introduces 1 alert** when merging 8181910c5de5986edcc0917a58a0887dbcf37c78 into f0598fd2289115c46cbc8bf59c38409a6a89b8a0 - [view on LGTM.com](https://lgtm.com/projects/g/uktrade/data-hub-frontend/rev/pr-eefbece1ace2b61591fcf178a96237c0128ed10c)

**new alerts:**

* 1 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by jim68000** [comment]
This pull request **introduces 1 alert** when merging 5afdd31fbbc1902f228dd61e2f527cee189da99e into 032029eeb9099031cb2a46a5ec1edfec7e99112c - [view on LGTM.com](https://lgtm.com/projects/g/uktrade/data-hub-frontend/rev/pr-3a9869f43ddafc048dacb0fb3c2e9105b2b87fcb)

**new alerts:**

* 1 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by jim68000** [comment]
This pull request **introduces 1 alert** when merging dff1654c2322096b5dea9bb10b58edd860a7b93e into 755b29a934f464a6df29f7699c05c911bb55b5ef - [view on LGTM.com](https://lgtm.com/projects/g/uktrade/data-hub-frontend/rev/pr-b283ddd818b57dcf2f0db199fd9a56f5c30c925a)

**new alerts:**

* 1 for Useless conditional

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: src/apps/interactions/services/interaction-options.js [hunk 1]
```diff
@@ -88,12 +88,35 @@ async function getServiceOptions (req, res, createdOn) {
...
 }
 
+function getExportContext (kind) {
+  if (kind === 'interaction') {
+    return 'export_interaction'
+  }
+  return 'export_service_delivery'
+}
+
+function getOtherContext (kind) {
```
Comment (yeahfro) [inline src/apps/interactions/services/interaction-options.js]: This is not covered by unit tests

File: src/apps/interactions/services/interaction-options.js [hunk 2]
```diff
@@ -88,12 +88,35 @@ async function getServiceOptions (req, res, createdOn) {
...
   if (get(res, 'locals.investment.id') || get(res.locals, 'interaction.investment_project.id')) {
     return 'investment_project_interaction'
   }
 
-  return snakeCase(get(req.params, 'kind', 'interaction'))
+  switch (theme) {
+    case 'export':
+      return getExportContext(kind)
+    case 'other':
+      return getOtherContext(kind)
```
Comment (yeahfro) [inline src/apps/interactions/services/interaction-options.js]: L 116 and L 118 are not covered by unit tests

File: src/apps/interactions/services/interaction-options.js [hunk 3]
```diff
@@ -88,12 +88,35 @@ async function getServiceOptions (req, res, createdOn) {
...
+
 function getContextForInteraction (req, res) {
+  const { theme = {}, kind } = req.params
+
   if (get(res, 'locals.investment.id') || get(res.locals, 'interaction.investment_project.id')) {
     return 'investment_project_interaction'
   }
 
-  return snakeCase(get(req.params, 'kind', 'interaction'))
+  switch (theme) {
```
Comment (yeahfro) [inline src/apps/interactions/services/interaction-options.js]: Something to think about instead of this `switch` is an `object`:

```
const getContext = {
  'export': (kind) => kind === 'interaction' ? 'export_interaction' : 'export_service_delivery',
  'other': (kind) => kind === 'interaction' ? 'other_interaction' : 'other_service_delivery',
}
```

Which is then invoked by:

`return getContext[theme] ? getContext[theme](kind) : 'investment_interaction'`

File: src/apps/interactions/controllers/edit.js [hunk 4]
```diff
@@ -41,6 +42,10 @@ async function buildForm (req, res, interactionId) {
     buttonText: interactionId ? 'Save and return' : `Add ${lowerCase(req.params.kind)}`,
   }
 
+  if (req.params.kind !== 'service-delivery' && req.params.kind !== 'interaction') {
+    res.redirect('/404')
```
Comment (yeahfro) [inline src/apps/interactions/controllers/edit.js]: This line is not covered

File: src/apps/interactions/services/interaction-options.js (hunk 1 above)
Comment (ian-leggett) [reply src/apps/interactions/services/interaction-options.js]: Do we need to test these as none of the functions inside this file are tested as we only care about the surface level. I think we only care about whats exported and that is `getInteractionOptions`

Comment (ian-leggett) [reply src/apps/interactions/services/interaction-options.js]: Actually changes below will remove these anyway

File: src/apps/interactions/controllers/edit.js (hunk 4 above)
Comment (ian-leggett) [reply src/apps/interactions/controllers/edit.js]: Just updated a test for this scenario

File: src/apps/interactions/services/interaction-options.js (hunk 3 above)
Comment (ian-leggett) [reply src/apps/interactions/services/interaction-options.js]: This is nice much nicer than switch, just been updated

File: src/apps/interactions/services/interaction-options.js (hunk 2 above)
Comment (ian-leggett) [reply src/apps/interactions/services/interaction-options.js]: We no longer need this now

File: src/apps/interactions/controllers/create.js [hunk 5]
```diff
@@ -2,17 +2,54 @@ const { get, kebabCase } = require('lodash')
 const { kindForm } = require('../macros')
 
 function postCreate (req, res, next) {
-  const kind = req.body.kind || null
+  let theme
```
[the remark this answers is not in the document]
Comment (ian-leggett) [reply src/apps/interactions/controllers/create.js]: Much better thanks 👍 All updated

File: src/apps/interactions/controllers/create.js [hunk 6]
```diff
@@ -1,18 +1,53 @@
...
+  },
+  other: {
+    kind: 'interaction',
+    theme: 'investment',
+  },
+}
+
 function postCreate (req, res, next) {
-  const kind = req.body.kind || null
+  let kindType = (req.body.kind_export || req.body.kind_other) || null
```
Comment (yeahfro) [inline src/apps/interactions/controllers/create.js]: This could be `const`

File: src/apps/interactions/controllers/create.js [hunk 7]
```diff
@@ -1,18 +1,53 @@
...
+
+  const { kind, theme } = kindLookup[kindType] || kindLookup.other
```
Comment (yeahfro) [inline src/apps/interactions/controllers/create.js]: My preference is to declare variables when I need them which would be after your error checks.

File: src/apps/interactions/services/interaction-options.js (hunk 1 above)
Comment (yeahfro) [reply src/apps/interactions/services/interaction-options.js]: I agree that we do not test internals but would expect scenarios could cover all lines by using the public function as the entry point.

File: src/apps/interactions/services/interaction-options.js [hunk 8]
```diff
@@ -89,11 +89,17 @@ async function getServiceOptions (req, res, createdOn) {
...
+  const { theme = {}, kind } = req.params
+
   if (get(res, 'locals.investment.id') || get(res.locals, 'interaction.investment_project.id')) {
     return 'investment_project_interaction'
   }
 
-  return snakeCase(get(req.params, 'kind', 'interaction'))
+  const getContext = {
+    'export': (kind) => kind === 'interaction' ? 'export_interaction' : 'export_service_delivery',
+    'other': (kind) => kind === 'interaction' ? 'other_interaction' : 'other_service_delivery',
```
Comment (yeahfro) [inline src/apps/interactions/services/interaction-options.js]: If this line is removed then all tests still pass.

File: src/apps/interactions/controllers/create.js (hunk 6 above)
Comment (ian-leggett) [reply src/apps/interactions/controllers/create.js]: This is a `const` now

File: src/apps/interactions/controllers/create.js (hunk 7 above)
Comment (ian-leggett) [reply src/apps/interactions/controllers/create.js]: Good spot, I have moved this now

File: src/apps/interactions/services/interaction-options.js (hunk 8 above)
Comment (ian-leggett) [reply src/apps/interactions/services/interaction-options.js]: Just updated the test, if you remove this line now it fails 👍

**Comment by paulgain** [comment]
Looks good @ian-leggett  
At some point in the future you should increase the coverage for repos.js within Interactions: https://43995-83823675-gh.circle-artifacts.com/0/root/data-hub-frontend/coverage/lcov-report/src/apps/interactions/index.html
